# 02 · Download and preprocess AERONET AOD Level 1.5 (Version 3), daily averages

**What this notebook does** — the same three steps as notebook 01, for the Level 1.5 product:

1. **Download** the NASA AERONET bulk archive (`.tar.gz`) into `DATA/RAW/` if it is not already there, extract it, and delete the archive.
2. **Read** every station file (`*.lev15`, one per AERONET site) and **merge** them into a single `pandas.DataFrame`.
3. **Export** one agent-ready Parquet table to `DATA/PROCESSED/`.

**Dataset**: AERONET **AOD Level 1.5** (cloud-screened and quality-controlled, but *not* finally calibrated), **Version 3**, **Daily averages**, all sites —
`https://aeronet.gsfc.nasa.gov/data_push/V3/AOD/AOD_Level15_Daily_V3.tar.gz` (~127 MB archive → ~670 MB extracted, ~1,613 sites, one row per site × day).

**Level 1.5 vs Level 2.0** (Giles et al. 2019): Level 1.5 is the near-real-time product. It has the same automatic cloud screening and instrument checks as
Level 2.0 but only the *pre-field* calibration, so it reaches the present day (Level 2.0 ends when the last post-field calibration was applied), includes
instruments that never get a final calibration, and can carry a calibration-drift bias (up to about +0.02 in AOD after a long deployment). The file header says
it: *"these data may not have final calibration applied. These data may change."* The columns are identical to Level 2.0; only `Data_Quality_Level` reads `lev15`.


In [1]:
from pathlib import Path
import re
import tarfile
import time

import numpy as np
import pandas as pd
import requests
from tqdm.auto import tqdm

/home/ubuntu/miniconda3/lib/python3.13/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [2]:
ARCHIVE = "AOD_Level15_Daily_V3.tar.gz"
URL = "https://aeronet.gsfc.nasa.gov/data_push/V3/AOD/" + ARCHIVE

In [3]:
PROJECT_ROOT = Path.cwd().resolve()
if PROJECT_ROOT.name == "process_data":          # this notebook lives in notebooks/process_data/
    PROJECT_ROOT = PROJECT_ROOT.parent.parent

In [4]:
RAW_DIR       = PROJECT_ROOT / "DATA" / "RAW"
PROCESSED_DIR = PROJECT_ROOT / "DATA" / "PROCESSED"
RAW_DIR.mkdir(parents=True, exist_ok=True)
PROCESSED_DIR.mkdir(parents=True, exist_ok=True)

In [5]:
TARBALL     = RAW_DIR / ARCHIVE                                  # deleted after extraction
EXTRACT_DIR = RAW_DIR / ARCHIVE.replace(".tar.gz", "")           # station files land here
OUT_PARQUET = PROCESSED_DIR / "AERONET_AOD_L15_Daily_V3.parquet"

In [6]:
print(f"URL          : {URL}")
print(f"PROJECT_ROOT : {PROJECT_ROOT}")
print(f"EXTRACT_DIR  : {EXTRACT_DIR}")
print(f"OUT_PARQUET  : {OUT_PARQUET}")

URL          : https://aeronet.gsfc.nasa.gov/data_push/V3/AOD/AOD_Level15_Daily_V3.tar.gz
PROJECT_ROOT : /home/ubuntu/ai_for_aeronet
EXTRACT_DIR  : /home/ubuntu/ai_for_aeronet/DATA/RAW/AOD_Level15_Daily_V3
OUT_PARQUET  : /home/ubuntu/ai_for_aeronet/DATA/PROCESSED/AERONET_AOD_L15_Daily_V3.parquet


## 1 · Download (if needed) → extract → delete the archive


In [7]:
def download(url: str, dest: Path, chunk_size: int = 1 << 20) -> None:
    tmp = dest.with_name(dest.name + ".part")
    with requests.get(url, stream=True, timeout=120) as r:
        r.raise_for_status()
        total = int(r.headers.get("Content-Length", 0)) or None
        with open(tmp, "wb") as fh, tqdm(total=total, unit="B", unit_scale=True, desc=dest.name) as bar:
            for part in r.iter_content(chunk_size):
                fh.write(part)
                bar.update(len(part))
    tmp.replace(dest)


def extract_tar_gz(archive: Path, dest: Path) -> None:
    dest.mkdir(parents=True, exist_ok=True)
    with tarfile.open(archive, "r:gz") as tf:
        try:
            tf.extractall(dest, filter="data")      
        except TypeError:                    
            tf.extractall(dest)


station_files = sorted(EXTRACT_DIR.rglob("*.lev15")) if EXTRACT_DIR.exists() else []

if station_files:
    print(f"Already extracted: {len(station_files)} station files in {EXTRACT_DIR}  →  skipping download.")
else:
    if TARBALL.exists():
        print(f"Found {TARBALL.name} ({TARBALL.stat().st_size / 1e6:,.0f} MB)  →  skipping download.")
    else:
        print(f"Downloading {URL}")
        t0 = time.time()
        download(URL, TARBALL)
        print(f"  done in {time.time() - t0:,.0f} s  ({TARBALL.stat().st_size / 1e6:,.0f} MB)")

    print(f"Extracting → {EXTRACT_DIR}")
    extract_tar_gz(TARBALL, EXTRACT_DIR)
    TARBALL.unlink()
    print(f"Deleted {TARBALL.name}")
    station_files = sorted(EXTRACT_DIR.rglob("*.lev15"))

print(f"{len(station_files):,} station files ready")

Found AOD_Level15_Daily_V3.tar.gz (127 MB)  →  skipping download.
Extracting → /home/ubuntu/ai_for_aeronet/DATA/RAW/AOD_Level15_Daily_V3


Deleted AOD_Level15_Daily_V3.tar.gz
1,613 station files ready


## 2 · Read every station file and merge

Each `*.lev15` file is one AERONET site, laid out exactly like the Level 2.0 files of notebook 01:

```
line 1  AERONET Version 3
line 2  <site name>
line 3  Version 3: AOD Level 1.5
line 4  The following data are cloud cleared and quality controls have been applied but these data may not have final calibration applied.  These data may change.
line 5  Contact: PI=<name>; PI Email=<email>
line 6  Daily Averages,UNITS can be found at,,, https://aeronet.gsfc.nasa.gov/new_web/units.html
line 7  <CSV header>
line 8+ <CSV rows>, missing values are -999
```


In [8]:
MISSING = -999.0

def read_station(path: Path) -> pd.DataFrame:
    # Parse one AERONET .lev15 file -> DataFrame (raw AERONET column names, -999 -> NaN,
    # plus `site`, `PI`, `PI_Email` taken from the file header).
    with open(path, encoding="latin-1") as fh:
        header = [next(fh).rstrip("\n") for _ in range(6)]
    site = header[1].strip()
    m = re.search(r"PI=(.*?);\s*PI Email=(.*)", header[4])
    pi, pi_email = (m.group(1).strip(), m.group(2).strip()) if m else (None, None)

    df = pd.read_csv(path, skiprows=6, encoding="latin-1", low_memory=False)

    # -999 → NaN in every numeric column (count columns never contain -999, so they stay integer)
    num = df.select_dtypes("number").columns
    df[num] = df[num].mask(df[num] == MISSING)

    df.insert(0, "site", site)
    df["PI"] = pi
    df["PI_Email"] = pi_email
    return df

In [9]:
frames, failures = [], []
t0 = time.time()
for p in tqdm(station_files, desc="Reading stations"):
    try:
        frames.append(read_station(p))
    except Exception as exc:            
        failures.append((p.name, repr(exc)))

aeronet_raw = pd.concat(frames, ignore_index=True)
del frames

print(f"Merged {len(station_files) - len(failures):,} stations → {aeronet_raw.shape[0]:,} rows × {aeronet_raw.shape[1]} columns "
      f"in {time.time() - t0:,.0f} s")
if failures:
    print(f"\n{len(failures)} file(s) failed to parse:")
    for name, err in failures:
        print(f"  {name}: {err}")

Reading stations:   0%|          | 0/1613 [00:00<?, ?it/s]

Reading stations:   0%|          | 3/1613 [00:00<00:56, 28.53it/s]

Reading stations:   1%|          | 9/1613 [00:00<00:35, 44.99it/s]

Reading stations:   1%|          | 15/1613 [00:00<00:34, 46.78it/s]

Reading stations:   1%|          | 20/1613 [00:00<00:50, 31.27it/s]

Reading stations:   2%|▏         | 26/1613 [00:00<00:42, 37.10it/s]

Reading stations:   2%|▏         | 31/1613 [00:00<00:40, 39.27it/s]

Reading stations:   2%|▏         | 36/1613 [00:00<00:37, 41.94it/s]

Reading stations:   3%|▎         | 41/1613 [00:01<00:38, 40.57it/s]

Reading stations:   3%|▎         | 46/1613 [00:01<00:40, 38.49it/s]

Reading stations:   3%|▎         | 51/1613 [00:01<00:40, 38.49it/s]

Reading stations:   3%|▎         | 56/1613 [00:01<00:42, 36.40it/s]

Reading stations:   4%|▎         | 60/1613 [00:01<00:43, 35.52it/s]

Reading stations:   4%|▍         | 65/1613 [00:01<00:41, 37.12it/s]

Reading stations:   4%|▍         | 71/1613 [00:01<00:37, 41.12it/s]

Reading stations:   5%|▍         | 77/1613 [00:01<00:39, 39.04it/s]

Reading stations:   5%|▌         | 81/1613 [00:02<00:45, 33.74it/s]

Reading stations:   5%|▌         | 85/1613 [00:02<00:48, 31.28it/s]

Reading stations:   6%|▌         | 89/1613 [00:02<00:51, 29.69it/s]

Reading stations:   6%|▌         | 93/1613 [00:02<00:56, 27.02it/s]

Reading stations:   6%|▌         | 96/1613 [00:02<01:02, 24.26it/s]

Reading stations:   6%|▌         | 99/1613 [00:02<01:02, 24.23it/s]

Reading stations:   6%|▋         | 102/1613 [00:03<01:08, 22.04it/s]

Reading stations:   7%|▋         | 106/1613 [00:03<01:06, 22.77it/s]

Reading stations:   7%|▋         | 109/1613 [00:03<01:02, 24.00it/s]

Reading stations:   7%|▋         | 112/1613 [00:03<01:05, 23.04it/s]

Reading stations:   7%|▋         | 115/1613 [00:03<01:08, 22.02it/s]

Reading stations:   7%|▋         | 118/1613 [00:03<01:27, 17.08it/s]

Reading stations:   7%|▋         | 120/1613 [00:04<01:27, 17.01it/s]

Reading stations:   8%|▊         | 123/1613 [00:04<01:18, 18.87it/s]

Reading stations:   8%|▊         | 126/1613 [00:04<01:17, 19.28it/s]

Reading stations:   8%|▊         | 129/1613 [00:04<01:18, 19.00it/s]

Reading stations:   8%|▊         | 131/1613 [00:04<01:19, 18.59it/s]

Reading stations:   8%|▊         | 134/1613 [00:04<01:14, 19.77it/s]

Reading stations:   8%|▊         | 137/1613 [00:04<01:15, 19.46it/s]

Reading stations:   9%|▊         | 139/1613 [00:05<01:26, 17.05it/s]

Reading stations:   9%|▉         | 144/1613 [00:05<01:14, 19.63it/s]

Reading stations:   9%|▉         | 148/1613 [00:05<01:03, 22.96it/s]

Reading stations:   9%|▉         | 151/1613 [00:05<01:01, 23.78it/s]

Reading stations:  10%|▉         | 154/1613 [00:05<01:07, 21.56it/s]

Reading stations:  10%|▉         | 157/1613 [00:05<01:02, 23.36it/s]

Reading stations:  10%|█         | 162/1613 [00:05<00:50, 28.53it/s]

Reading stations:  10%|█         | 167/1613 [00:06<00:45, 32.03it/s]

Reading stations:  11%|█         | 171/1613 [00:06<00:50, 28.66it/s]

Reading stations:  11%|█         | 175/1613 [00:06<00:46, 31.03it/s]

Reading stations:  11%|█         | 179/1613 [00:06<00:45, 31.73it/s]

Reading stations:  11%|█▏        | 184/1613 [00:06<00:39, 35.74it/s]

Reading stations:  12%|█▏        | 188/1613 [00:06<00:40, 35.32it/s]

Reading stations:  12%|█▏        | 192/1613 [00:06<00:43, 32.91it/s]

Reading stations:  12%|█▏        | 196/1613 [00:06<00:45, 31.09it/s]

Reading stations:  13%|█▎        | 202/1613 [00:07<00:39, 35.89it/s]

Reading stations:  13%|█▎        | 206/1613 [00:07<00:44, 31.94it/s]

Reading stations:  13%|█▎        | 210/1613 [00:07<00:44, 31.33it/s]

Reading stations:  13%|█▎        | 214/1613 [00:07<00:44, 31.76it/s]

Reading stations:  14%|█▎        | 218/1613 [00:07<00:52, 26.43it/s]

Reading stations:  14%|█▎        | 221/1613 [00:07<00:55, 24.88it/s]

Reading stations:  14%|█▍        | 225/1613 [00:07<00:50, 27.51it/s]

Reading stations:  14%|█▍        | 229/1613 [00:08<00:46, 29.57it/s]

Reading stations:  14%|█▍        | 233/1613 [00:08<00:54, 25.37it/s]

Reading stations:  15%|█▍        | 236/1613 [00:08<00:56, 24.53it/s]

Reading stations:  15%|█▍        | 239/1613 [00:08<00:55, 24.56it/s]

Reading stations:  15%|█▌        | 242/1613 [00:08<00:59, 22.98it/s]

Reading stations:  15%|█▌        | 245/1613 [00:08<00:58, 23.57it/s]

Reading stations:  16%|█▌        | 251/1613 [00:08<00:43, 31.59it/s]

Reading stations:  16%|█▌        | 257/1613 [00:09<00:36, 37.44it/s]

Reading stations:  16%|█▌        | 262/1613 [00:09<00:33, 40.00it/s]

Reading stations:  17%|█▋        | 268/1613 [00:09<00:31, 43.33it/s]

Reading stations:  17%|█▋        | 273/1613 [00:09<00:30, 44.15it/s]

Reading stations:  17%|█▋        | 278/1613 [00:09<00:29, 44.63it/s]

Reading stations:  18%|█▊        | 283/1613 [00:09<00:43, 30.48it/s]

Reading stations:  18%|█▊        | 288/1613 [00:09<00:38, 34.46it/s]

Reading stations:  18%|█▊        | 293/1613 [00:10<00:48, 27.28it/s]

Reading stations:  18%|█▊        | 297/1613 [00:10<00:57, 22.92it/s]

Reading stations:  19%|█▊        | 301/1613 [00:10<00:50, 25.75it/s]

Reading stations:  19%|█▉        | 305/1613 [00:10<01:02, 20.87it/s]

Reading stations:  19%|█▉        | 309/1613 [00:10<00:56, 23.26it/s]

Reading stations:  19%|█▉        | 312/1613 [00:11<01:06, 19.54it/s]

Reading stations:  20%|█▉        | 315/1613 [00:11<01:10, 18.33it/s]

Reading stations:  20%|█▉        | 318/1613 [00:11<01:06, 19.35it/s]

Reading stations:  20%|█▉        | 321/1613 [00:11<01:02, 20.61it/s]

Reading stations:  20%|██        | 324/1613 [00:11<00:57, 22.48it/s]

Reading stations:  20%|██        | 327/1613 [00:11<00:55, 23.12it/s]

Reading stations:  20%|██        | 330/1613 [00:11<00:55, 22.98it/s]

Reading stations:  21%|██        | 333/1613 [00:12<00:56, 22.79it/s]

Reading stations:  21%|██        | 336/1613 [00:12<00:59, 21.52it/s]

Reading stations:  21%|██        | 339/1613 [00:12<01:07, 18.75it/s]

Reading stations:  21%|██        | 342/1613 [00:12<01:00, 20.92it/s]

Reading stations:  21%|██▏       | 345/1613 [00:12<00:56, 22.55it/s]

Reading stations:  22%|██▏       | 350/1613 [00:12<00:44, 28.08it/s]

Reading stations:  22%|██▏       | 353/1613 [00:12<00:45, 27.62it/s]

Reading stations:  22%|██▏       | 356/1613 [00:13<00:53, 23.55it/s]

Reading stations:  22%|██▏       | 359/1613 [00:13<00:59, 21.01it/s]

Reading stations:  22%|██▏       | 362/1613 [00:13<01:01, 20.35it/s]

Reading stations:  23%|██▎       | 366/1613 [00:13<00:52, 23.58it/s]

Reading stations:  23%|██▎       | 369/1613 [00:13<01:05, 18.96it/s]

Reading stations:  23%|██▎       | 372/1613 [00:13<01:08, 18.15it/s]

Reading stations:  23%|██▎       | 375/1613 [00:14<01:02, 19.67it/s]

Reading stations:  23%|██▎       | 378/1613 [00:14<00:57, 21.60it/s]

Reading stations:  24%|██▎       | 382/1613 [00:14<00:50, 24.54it/s]

Reading stations:  24%|██▍       | 387/1613 [00:14<00:41, 29.70it/s]

Reading stations:  24%|██▍       | 391/1613 [00:14<00:38, 31.69it/s]

Reading stations:  24%|██▍       | 395/1613 [00:14<00:40, 30.33it/s]

Reading stations:  25%|██▍       | 399/1613 [00:14<00:37, 32.20it/s]

Reading stations:  25%|██▌       | 404/1613 [00:14<00:36, 32.84it/s]

Reading stations:  25%|██▌       | 408/1613 [00:15<00:48, 24.68it/s]

Reading stations:  25%|██▌       | 411/1613 [00:15<00:49, 24.11it/s]

Reading stations:  26%|██▌       | 414/1613 [00:15<00:47, 25.34it/s]

Reading stations:  26%|██▌       | 418/1613 [00:15<00:43, 27.59it/s]

Reading stations:  26%|██▌       | 422/1613 [00:15<00:40, 29.43it/s]

Reading stations:  26%|██▋       | 426/1613 [00:15<00:45, 25.98it/s]

Reading stations:  27%|██▋       | 429/1613 [00:15<00:44, 26.57it/s]

Reading stations:  27%|██▋       | 434/1613 [00:16<00:38, 30.65it/s]

Reading stations:  27%|██▋       | 438/1613 [00:16<00:36, 31.97it/s]

Reading stations:  27%|██▋       | 442/1613 [00:16<00:35, 33.17it/s]

Reading stations:  28%|██▊       | 446/1613 [00:16<00:35, 32.81it/s]

Reading stations:  28%|██▊       | 450/1613 [00:16<00:36, 31.77it/s]

Reading stations:  28%|██▊       | 454/1613 [00:16<00:34, 33.18it/s]

Reading stations:  28%|██▊       | 458/1613 [00:16<00:42, 27.26it/s]

Reading stations:  29%|██▊       | 461/1613 [00:16<00:42, 27.28it/s]

Reading stations:  29%|██▉       | 465/1613 [00:17<00:38, 30.14it/s]

Reading stations:  29%|██▉       | 470/1613 [00:17<00:33, 33.90it/s]

Reading stations:  29%|██▉       | 474/1613 [00:17<00:35, 31.91it/s]

Reading stations:  30%|██▉       | 478/1613 [00:17<00:36, 31.19it/s]

Reading stations:  30%|██▉       | 482/1613 [00:17<00:36, 31.28it/s]

Reading stations:  30%|███       | 486/1613 [00:17<00:37, 29.87it/s]

Reading stations:  30%|███       | 490/1613 [00:17<00:36, 30.74it/s]

Reading stations:  31%|███       | 495/1613 [00:17<00:32, 34.42it/s]

Reading stations:  31%|███       | 499/1613 [00:18<00:31, 34.85it/s]

Reading stations:  31%|███       | 503/1613 [00:18<00:32, 34.53it/s]

Reading stations:  31%|███▏      | 508/1613 [00:18<00:28, 38.32it/s]

Reading stations:  32%|███▏      | 515/1613 [00:18<00:24, 44.94it/s]

Reading stations:  32%|███▏      | 520/1613 [00:18<00:28, 38.81it/s]

Reading stations:  33%|███▎      | 525/1613 [00:18<00:35, 31.01it/s]

Reading stations:  33%|███▎      | 530/1613 [00:18<00:34, 31.05it/s]

Reading stations:  33%|███▎      | 534/1613 [00:19<00:36, 29.43it/s]

Reading stations:  33%|███▎      | 538/1613 [00:19<00:34, 31.32it/s]

Reading stations:  34%|███▎      | 543/1613 [00:19<00:32, 33.09it/s]

Reading stations:  34%|███▍      | 547/1613 [00:19<00:33, 32.18it/s]

Reading stations:  34%|███▍      | 551/1613 [00:19<00:54, 19.63it/s]

Reading stations:  34%|███▍      | 554/1613 [00:20<00:51, 20.45it/s]

Reading stations:  35%|███▍      | 560/1613 [00:20<00:38, 27.34it/s]

Reading stations:  35%|███▍      | 564/1613 [00:20<00:36, 29.13it/s]

Reading stations:  35%|███▌      | 568/1613 [00:20<00:34, 30.11it/s]

Reading stations:  35%|███▌      | 572/1613 [00:20<00:36, 28.67it/s]

Reading stations:  36%|███▌      | 576/1613 [00:20<00:46, 22.50it/s]

Reading stations:  36%|███▌      | 579/1613 [00:20<00:44, 23.15it/s]

Reading stations:  36%|███▌      | 582/1613 [00:21<00:48, 21.13it/s]

Reading stations:  36%|███▋      | 585/1613 [00:21<00:45, 22.80it/s]

Reading stations:  37%|███▋      | 589/1613 [00:21<00:39, 26.05it/s]

Reading stations:  37%|███▋      | 592/1613 [00:21<00:41, 24.68it/s]

Reading stations:  37%|███▋      | 595/1613 [00:21<00:51, 19.62it/s]

Reading stations:  37%|███▋      | 598/1613 [00:21<00:55, 18.38it/s]

Reading stations:  37%|███▋      | 602/1613 [00:22<00:48, 20.91it/s]

Reading stations:  38%|███▊      | 606/1613 [00:22<00:41, 23.99it/s]

Reading stations:  38%|███▊      | 609/1613 [00:22<00:45, 22.24it/s]

Reading stations:  38%|███▊      | 612/1613 [00:22<00:47, 20.91it/s]

Reading stations:  38%|███▊      | 616/1613 [00:22<00:40, 24.47it/s]

Reading stations:  38%|███▊      | 620/1613 [00:22<00:36, 27.17it/s]

Reading stations:  39%|███▊      | 623/1613 [00:22<00:42, 23.26it/s]

Reading stations:  39%|███▉      | 629/1613 [00:22<00:31, 31.39it/s]

Reading stations:  39%|███▉      | 633/1613 [00:23<00:33, 29.39it/s]

Reading stations:  39%|███▉      | 637/1613 [00:23<00:32, 30.41it/s]

Reading stations:  40%|███▉      | 641/1613 [00:23<00:34, 28.49it/s]

Reading stations:  40%|███▉      | 645/1613 [00:23<00:31, 31.02it/s]

Reading stations:  40%|████      | 649/1613 [00:23<00:35, 26.89it/s]

Reading stations:  40%|████      | 653/1613 [00:23<00:32, 29.78it/s]

Reading stations:  41%|████      | 657/1613 [00:23<00:29, 31.90it/s]

Reading stations:  41%|████      | 661/1613 [00:24<00:31, 29.93it/s]

Reading stations:  41%|████      | 665/1613 [00:24<00:32, 29.36it/s]

Reading stations:  42%|████▏     | 671/1613 [00:24<00:26, 35.45it/s]

Reading stations:  42%|████▏     | 675/1613 [00:24<00:30, 30.40it/s]

Reading stations:  42%|████▏     | 679/1613 [00:24<00:29, 32.05it/s]

Reading stations:  42%|████▏     | 683/1613 [00:24<00:30, 30.52it/s]

Reading stations:  43%|████▎     | 687/1613 [00:24<00:28, 32.53it/s]

Reading stations:  43%|████▎     | 691/1613 [00:25<00:28, 32.87it/s]

Reading stations:  43%|████▎     | 695/1613 [00:25<00:26, 34.42it/s]

Reading stations:  43%|████▎     | 699/1613 [00:25<00:28, 32.04it/s]

Reading stations:  44%|████▎     | 703/1613 [00:25<00:27, 32.64it/s]

Reading stations:  44%|████▍     | 707/1613 [00:25<00:26, 33.59it/s]

Reading stations:  44%|████▍     | 711/1613 [00:25<00:28, 31.86it/s]

Reading stations:  44%|████▍     | 715/1613 [00:25<00:28, 31.45it/s]

Reading stations:  45%|████▍     | 719/1613 [00:25<00:31, 28.22it/s]

Reading stations:  45%|████▍     | 723/1613 [00:26<00:29, 29.74it/s]

Reading stations:  45%|████▌     | 727/1613 [00:26<00:31, 28.19it/s]

Reading stations:  45%|████▌     | 730/1613 [00:26<00:30, 28.50it/s]

Reading stations:  46%|████▌     | 734/1613 [00:26<00:32, 26.78it/s]

Reading stations:  46%|████▌     | 738/1613 [00:26<00:31, 27.98it/s]

Reading stations:  46%|████▌     | 742/1613 [00:26<00:28, 30.55it/s]

Reading stations:  46%|████▌     | 746/1613 [00:27<00:39, 21.82it/s]

Reading stations:  46%|████▋     | 749/1613 [00:27<00:46, 18.56it/s]

Reading stations:  47%|████▋     | 752/1613 [00:27<00:50, 17.12it/s]

Reading stations:  47%|████▋     | 755/1613 [00:27<00:48, 17.66it/s]

Reading stations:  47%|████▋     | 758/1613 [00:27<00:44, 19.16it/s]

Reading stations:  47%|████▋     | 761/1613 [00:27<00:45, 18.61it/s]

Reading stations:  47%|████▋     | 766/1613 [00:28<00:34, 24.38it/s]

Reading stations:  48%|████▊     | 770/1613 [00:28<00:32, 26.08it/s]

Reading stations:  48%|████▊     | 776/1613 [00:28<00:25, 32.54it/s]

Reading stations:  48%|████▊     | 782/1613 [00:28<00:23, 35.03it/s]

Reading stations:  49%|████▉     | 788/1613 [00:28<00:20, 39.35it/s]

Reading stations:  49%|████▉     | 793/1613 [00:28<00:19, 41.61it/s]

Reading stations:  49%|████▉     | 798/1613 [00:28<00:18, 43.63it/s]

Reading stations:  50%|████▉     | 803/1613 [00:28<00:24, 33.23it/s]

Reading stations:  50%|█████     | 807/1613 [00:29<00:25, 31.83it/s]

Reading stations:  50%|█████     | 812/1613 [00:29<00:23, 34.46it/s]

Reading stations:  51%|█████     | 816/1613 [00:29<00:43, 18.30it/s]

Reading stations:  51%|█████     | 819/1613 [00:29<00:41, 18.97it/s]

Reading stations:  51%|█████     | 824/1613 [00:29<00:33, 23.85it/s]

Reading stations:  51%|█████▏    | 828/1613 [00:30<00:30, 25.40it/s]

Reading stations:  52%|█████▏    | 833/1613 [00:30<00:26, 29.08it/s]

Reading stations:  52%|█████▏    | 837/1613 [00:30<00:28, 27.66it/s]

Reading stations:  52%|█████▏    | 841/1613 [00:30<00:25, 29.71it/s]

Reading stations:  52%|█████▏    | 845/1613 [00:30<00:26, 28.70it/s]

Reading stations:  53%|█████▎    | 849/1613 [00:30<00:26, 28.67it/s]

Reading stations:  53%|█████▎    | 853/1613 [00:30<00:26, 28.27it/s]

Reading stations:  53%|█████▎    | 856/1613 [00:31<00:27, 27.94it/s]

Reading stations:  53%|█████▎    | 859/1613 [00:31<00:31, 24.31it/s]

Reading stations:  53%|█████▎    | 862/1613 [00:31<00:43, 17.20it/s]

Reading stations:  54%|█████▎    | 865/1613 [00:31<00:50, 14.75it/s]

Reading stations:  54%|█████▍    | 868/1613 [00:31<00:45, 16.52it/s]

Reading stations:  54%|█████▍    | 871/1613 [00:32<00:39, 18.58it/s]

Reading stations:  54%|█████▍    | 874/1613 [00:32<00:40, 18.08it/s]

Reading stations:  54%|█████▍    | 878/1613 [00:32<00:33, 21.90it/s]

Reading stations:  55%|█████▍    | 883/1613 [00:32<00:27, 27.03it/s]

Reading stations:  55%|█████▍    | 887/1613 [00:32<00:29, 24.28it/s]

Reading stations:  55%|█████▌    | 891/1613 [00:32<00:30, 23.76it/s]

Reading stations:  56%|█████▌    | 897/1613 [00:32<00:24, 29.74it/s]

Reading stations:  56%|█████▌    | 902/1613 [00:33<00:20, 34.17it/s]

Reading stations:  56%|█████▌    | 906/1613 [00:33<00:20, 34.13it/s]

Reading stations:  56%|█████▋    | 910/1613 [00:33<00:22, 31.57it/s]

Reading stations:  57%|█████▋    | 916/1613 [00:33<00:19, 36.62it/s]

Reading stations:  57%|█████▋    | 920/1613 [00:33<00:24, 28.00it/s]

Reading stations:  57%|█████▋    | 925/1613 [00:33<00:23, 29.30it/s]

Reading stations:  58%|█████▊    | 929/1613 [00:33<00:21, 31.16it/s]

Reading stations:  58%|█████▊    | 933/1613 [00:34<00:23, 28.87it/s]

Reading stations:  58%|█████▊    | 937/1613 [00:34<00:22, 30.62it/s]

Reading stations:  58%|█████▊    | 941/1613 [00:34<00:24, 27.40it/s]

Reading stations:  59%|█████▊    | 945/1613 [00:34<00:23, 28.31it/s]

Reading stations:  59%|█████▉    | 950/1613 [00:34<00:21, 31.16it/s]

Reading stations:  59%|█████▉    | 954/1613 [00:34<00:22, 29.13it/s]

Reading stations:  59%|█████▉    | 958/1613 [00:35<00:25, 25.53it/s]

Reading stations:  60%|█████▉    | 964/1613 [00:35<00:19, 32.55it/s]

Reading stations:  60%|██████    | 969/1613 [00:35<00:19, 32.33it/s]

Reading stations:  60%|██████    | 974/1613 [00:35<00:17, 35.63it/s]

Reading stations:  61%|██████    | 978/1613 [00:35<00:21, 29.13it/s]

Reading stations:  61%|██████    | 982/1613 [00:35<00:20, 30.17it/s]

Reading stations:  61%|██████    | 986/1613 [00:35<00:19, 31.56it/s]

Reading stations:  61%|██████▏   | 990/1613 [00:36<00:22, 27.29it/s]

Reading stations:  62%|██████▏   | 997/1613 [00:36<00:17, 35.08it/s]

Reading stations:  62%|██████▏   | 1004/1613 [00:36<00:14, 41.33it/s]

Reading stations:  63%|██████▎   | 1009/1613 [00:36<00:14, 42.08it/s]

Reading stations:  63%|██████▎   | 1015/1613 [00:36<00:13, 45.99it/s]

Reading stations:  63%|██████▎   | 1020/1613 [00:36<00:20, 28.97it/s]

Reading stations:  63%|██████▎   | 1024/1613 [00:37<00:24, 24.13it/s]

Reading stations:  64%|██████▎   | 1028/1613 [00:37<00:27, 21.42it/s]

Reading stations:  64%|██████▍   | 1031/1613 [00:37<00:27, 21.47it/s]

Reading stations:  64%|██████▍   | 1034/1613 [00:37<00:27, 21.25it/s]

Reading stations:  64%|██████▍   | 1038/1613 [00:37<00:23, 24.56it/s]

Reading stations:  65%|██████▍   | 1041/1613 [00:37<00:26, 21.73it/s]

Reading stations:  65%|██████▍   | 1045/1613 [00:38<00:23, 24.53it/s]

Reading stations:  65%|██████▍   | 1048/1613 [00:38<00:22, 25.41it/s]

Reading stations:  65%|██████▌   | 1051/1613 [00:38<00:24, 22.61it/s]

Reading stations:  65%|██████▌   | 1054/1613 [00:38<00:23, 24.09it/s]

Reading stations:  66%|██████▌   | 1057/1613 [00:38<00:23, 24.00it/s]

Reading stations:  66%|██████▌   | 1061/1613 [00:38<00:20, 27.24it/s]

Reading stations:  66%|██████▌   | 1064/1613 [00:38<00:24, 22.06it/s]

Reading stations:  66%|██████▌   | 1067/1613 [00:39<00:26, 20.64it/s]

Reading stations:  66%|██████▋   | 1070/1613 [00:39<00:27, 19.46it/s]

Reading stations:  67%|██████▋   | 1073/1613 [00:39<00:25, 21.05it/s]

Reading stations:  67%|██████▋   | 1076/1613 [00:39<00:26, 20.23it/s]

Reading stations:  67%|██████▋   | 1079/1613 [00:39<00:27, 19.56it/s]

Reading stations:  67%|██████▋   | 1082/1613 [00:40<00:52, 10.10it/s]

Reading stations:  67%|██████▋   | 1085/1613 [00:40<00:43, 12.00it/s]

Reading stations:  67%|██████▋   | 1088/1613 [00:40<00:36, 14.27it/s]

Reading stations:  68%|██████▊   | 1091/1613 [00:40<00:33, 15.63it/s]

Reading stations:  68%|██████▊   | 1094/1613 [00:40<00:31, 16.67it/s]

Reading stations:  68%|██████▊   | 1097/1613 [00:40<00:27, 18.81it/s]

Reading stations:  68%|██████▊   | 1101/1613 [00:41<00:22, 22.31it/s]

Reading stations:  68%|██████▊   | 1104/1613 [00:41<00:21, 23.43it/s]

Reading stations:  69%|██████▊   | 1108/1613 [00:41<00:19, 25.42it/s]

Reading stations:  69%|██████▉   | 1112/1613 [00:41<00:17, 28.19it/s]

Reading stations:  69%|██████▉   | 1115/1613 [00:41<00:19, 26.10it/s]

Reading stations:  69%|██████▉   | 1118/1613 [00:41<00:19, 25.04it/s]

Reading stations:  69%|██████▉   | 1121/1613 [00:41<00:21, 22.68it/s]

Reading stations:  70%|██████▉   | 1125/1613 [00:41<00:19, 25.48it/s]

Reading stations:  70%|██████▉   | 1128/1613 [00:42<00:18, 26.52it/s]

Reading stations:  70%|███████   | 1133/1613 [00:42<00:16, 28.57it/s]

Reading stations:  70%|███████   | 1136/1613 [00:42<00:17, 27.77it/s]

Reading stations:  71%|███████   | 1140/1613 [00:42<00:16, 29.10it/s]

Reading stations:  71%|███████   | 1143/1613 [00:42<00:16, 28.84it/s]

Reading stations:  71%|███████   | 1148/1613 [00:42<00:16, 27.36it/s]

Reading stations:  71%|███████▏  | 1152/1613 [00:42<00:15, 29.27it/s]

Reading stations:  72%|███████▏  | 1156/1613 [00:43<00:14, 31.76it/s]

Reading stations:  72%|███████▏  | 1160/1613 [00:43<00:14, 30.66it/s]

Reading stations:  72%|███████▏  | 1165/1613 [00:43<00:13, 33.63it/s]

Reading stations:  72%|███████▏  | 1169/1613 [00:43<00:15, 28.20it/s]

Reading stations:  73%|███████▎  | 1174/1613 [00:43<00:13, 32.06it/s]

Reading stations:  73%|███████▎  | 1178/1613 [00:43<00:13, 33.04it/s]

Reading stations:  73%|███████▎  | 1182/1613 [00:43<00:14, 29.92it/s]

Reading stations:  74%|███████▎  | 1186/1613 [00:43<00:13, 31.61it/s]

Reading stations:  74%|███████▍  | 1190/1613 [00:44<00:20, 20.88it/s]

Reading stations:  74%|███████▍  | 1193/1613 [00:44<00:21, 19.63it/s]

Reading stations:  74%|███████▍  | 1197/1613 [00:44<00:18, 22.14it/s]

Reading stations:  74%|███████▍  | 1200/1613 [00:44<00:21, 19.26it/s]

Reading stations:  75%|███████▍  | 1204/1613 [00:44<00:18, 22.25it/s]

Reading stations:  75%|███████▍  | 1207/1613 [00:45<00:18, 21.60it/s]

Reading stations:  75%|███████▌  | 1210/1613 [00:45<00:20, 19.83it/s]

Reading stations:  75%|███████▌  | 1213/1613 [00:45<00:18, 21.63it/s]

Reading stations:  75%|███████▌  | 1217/1613 [00:45<00:15, 25.52it/s]

Reading stations:  76%|███████▌  | 1220/1613 [00:45<00:17, 22.78it/s]

Reading stations:  76%|███████▌  | 1224/1613 [00:45<00:14, 26.28it/s]

Reading stations:  76%|███████▋  | 1230/1613 [00:45<00:11, 32.11it/s]

Reading stations:  77%|███████▋  | 1234/1613 [00:46<00:11, 32.32it/s]

Reading stations:  77%|███████▋  | 1238/1613 [00:46<00:12, 30.93it/s]

Reading stations:  77%|███████▋  | 1242/1613 [00:46<00:12, 30.52it/s]

Reading stations:  77%|███████▋  | 1246/1613 [00:46<00:12, 29.16it/s]

Reading stations:  77%|███████▋  | 1249/1613 [00:46<00:15, 24.26it/s]

Reading stations:  78%|███████▊  | 1252/1613 [00:46<00:16, 22.52it/s]

Reading stations:  78%|███████▊  | 1255/1613 [00:46<00:17, 20.65it/s]

Reading stations:  78%|███████▊  | 1258/1613 [00:47<00:16, 21.76it/s]

Reading stations:  78%|███████▊  | 1261/1613 [00:47<00:15, 22.93it/s]

Reading stations:  78%|███████▊  | 1264/1613 [00:47<00:15, 22.85it/s]

Reading stations:  79%|███████▊  | 1267/1613 [00:47<00:14, 23.24it/s]

Reading stations:  79%|███████▊  | 1270/1613 [00:47<00:13, 24.73it/s]

Reading stations:  79%|███████▉  | 1273/1613 [00:47<00:18, 18.25it/s]

Reading stations:  79%|███████▉  | 1277/1613 [00:47<00:15, 21.42it/s]

Reading stations:  79%|███████▉  | 1280/1613 [00:48<00:16, 20.37it/s]

Reading stations:  80%|███████▉  | 1284/1613 [00:48<00:13, 23.91it/s]

Reading stations:  80%|███████▉  | 1288/1613 [00:48<00:11, 27.16it/s]

Reading stations:  80%|████████  | 1291/1613 [00:48<00:12, 26.51it/s]

Reading stations:  80%|████████  | 1296/1613 [00:48<00:10, 31.45it/s]

Reading stations:  81%|████████  | 1301/1613 [00:48<00:08, 35.55it/s]

Reading stations:  81%|████████  | 1305/1613 [00:48<00:11, 26.40it/s]

Reading stations:  81%|████████  | 1309/1613 [00:49<00:10, 28.02it/s]

Reading stations:  82%|████████▏ | 1315/1613 [00:49<00:08, 33.78it/s]

Reading stations:  82%|████████▏ | 1319/1613 [00:49<00:08, 34.62it/s]

Reading stations:  82%|████████▏ | 1323/1613 [00:49<00:08, 32.94it/s]

Reading stations:  82%|████████▏ | 1327/1613 [00:49<00:08, 32.36it/s]

Reading stations:  83%|████████▎ | 1331/1613 [00:49<00:08, 33.00it/s]

Reading stations:  83%|████████▎ | 1335/1613 [00:49<00:08, 32.19it/s]

Reading stations:  83%|████████▎ | 1340/1613 [00:49<00:07, 35.69it/s]

Reading stations:  83%|████████▎ | 1344/1613 [00:50<00:07, 36.06it/s]

Reading stations:  84%|████████▎ | 1348/1613 [00:50<00:08, 30.19it/s]

Reading stations:  84%|████████▍ | 1352/1613 [00:50<00:08, 30.37it/s]

Reading stations:  84%|████████▍ | 1356/1613 [00:50<00:09, 27.12it/s]

Reading stations:  84%|████████▍ | 1359/1613 [00:50<00:09, 26.08it/s]

Reading stations:  84%|████████▍ | 1362/1613 [00:50<00:09, 26.42it/s]

Reading stations:  85%|████████▍ | 1365/1613 [00:50<00:10, 22.82it/s]

Reading stations:  85%|████████▍ | 1370/1613 [00:51<00:08, 28.18it/s]

Reading stations:  85%|████████▌ | 1374/1613 [00:51<00:08, 28.69it/s]

Reading stations:  85%|████████▌ | 1378/1613 [00:51<00:09, 26.09it/s]

Reading stations:  86%|████████▌ | 1381/1613 [00:51<00:08, 26.55it/s]

Reading stations:  86%|████████▌ | 1385/1613 [00:51<00:08, 26.14it/s]

Reading stations:  86%|████████▌ | 1389/1613 [00:51<00:07, 28.37it/s]

Reading stations:  86%|████████▋ | 1392/1613 [00:51<00:07, 28.01it/s]

Reading stations:  86%|████████▋ | 1395/1613 [00:52<00:16, 13.08it/s]

Reading stations:  87%|████████▋ | 1398/1613 [00:52<00:14, 15.23it/s]

Reading stations:  87%|████████▋ | 1403/1613 [00:52<00:10, 20.75it/s]

Reading stations:  87%|████████▋ | 1407/1613 [00:52<00:09, 22.68it/s]

Reading stations:  88%|████████▊ | 1412/1613 [00:52<00:07, 26.38it/s]

Reading stations:  88%|████████▊ | 1416/1613 [00:53<00:06, 28.43it/s]

Reading stations:  88%|████████▊ | 1422/1613 [00:53<00:05, 34.91it/s]

Reading stations:  88%|████████▊ | 1426/1613 [00:53<00:05, 31.28it/s]

Reading stations:  89%|████████▊ | 1430/1613 [00:53<00:06, 29.28it/s]

Reading stations:  89%|████████▉ | 1434/1613 [00:53<00:05, 30.03it/s]

Reading stations:  89%|████████▉ | 1438/1613 [00:53<00:06, 27.25it/s]

Reading stations:  89%|████████▉ | 1441/1613 [00:53<00:07, 24.15it/s]

Reading stations:  90%|████████▉ | 1446/1613 [00:54<00:05, 28.59it/s]

Reading stations:  90%|████████▉ | 1450/1613 [00:54<00:06, 26.00it/s]

Reading stations:  90%|█████████ | 1455/1613 [00:54<00:05, 30.98it/s]

Reading stations:  90%|█████████ | 1459/1613 [00:54<00:05, 29.06it/s]

Reading stations:  91%|█████████ | 1463/1613 [00:54<00:05, 28.52it/s]

Reading stations:  91%|█████████ | 1467/1613 [00:54<00:05, 27.51it/s]

Reading stations:  91%|█████████ | 1471/1613 [00:55<00:05, 26.02it/s]

Reading stations:  91%|█████████▏| 1474/1613 [00:55<00:05, 24.07it/s]

Reading stations:  92%|█████████▏| 1478/1613 [00:55<00:05, 23.55it/s]

Reading stations:  92%|█████████▏| 1483/1613 [00:55<00:04, 28.37it/s]

Reading stations:  92%|█████████▏| 1487/1613 [00:55<00:04, 27.10it/s]

Reading stations:  92%|█████████▏| 1491/1613 [00:55<00:04, 28.48it/s]

Reading stations:  93%|█████████▎| 1494/1613 [00:55<00:04, 28.48it/s]

Reading stations:  93%|█████████▎| 1499/1613 [00:55<00:03, 33.15it/s]

Reading stations:  93%|█████████▎| 1503/1613 [00:56<00:03, 31.88it/s]

Reading stations:  93%|█████████▎| 1507/1613 [00:56<00:03, 31.60it/s]

Reading stations:  94%|█████████▎| 1511/1613 [00:56<00:03, 31.51it/s]

Reading stations:  94%|█████████▍| 1515/1613 [00:56<00:03, 27.93it/s]

Reading stations:  94%|█████████▍| 1519/1613 [00:56<00:03, 29.93it/s]

Reading stations:  94%|█████████▍| 1523/1613 [00:56<00:03, 25.91it/s]

Reading stations:  95%|█████████▍| 1526/1613 [00:56<00:03, 25.94it/s]

Reading stations:  95%|█████████▍| 1531/1613 [00:57<00:02, 31.05it/s]

Reading stations:  95%|█████████▌| 1535/1613 [00:57<00:03, 25.01it/s]

Reading stations:  95%|█████████▌| 1539/1613 [00:57<00:02, 25.72it/s]

Reading stations:  96%|█████████▌| 1542/1613 [00:57<00:02, 25.21it/s]

Reading stations:  96%|█████████▌| 1545/1613 [00:57<00:02, 23.08it/s]

Reading stations:  96%|█████████▌| 1549/1613 [00:57<00:02, 26.51it/s]

Reading stations:  96%|█████████▋| 1553/1613 [00:57<00:02, 29.15it/s]

Reading stations:  97%|█████████▋| 1557/1613 [00:58<00:01, 30.35it/s]

Reading stations:  97%|█████████▋| 1561/1613 [00:58<00:01, 29.08it/s]

Reading stations:  97%|█████████▋| 1565/1613 [00:58<00:01, 27.50it/s]

Reading stations:  97%|█████████▋| 1570/1613 [00:58<00:01, 31.45it/s]

Reading stations:  98%|█████████▊| 1574/1613 [00:58<00:01, 29.04it/s]

Reading stations:  98%|█████████▊| 1579/1613 [00:58<00:01, 33.12it/s]

Reading stations:  98%|█████████▊| 1583/1613 [00:58<00:01, 29.05it/s]

Reading stations:  98%|█████████▊| 1587/1613 [00:59<00:00, 26.96it/s]

Reading stations:  99%|█████████▊| 1591/1613 [00:59<00:00, 29.06it/s]

Reading stations:  99%|█████████▉| 1595/1613 [00:59<00:00, 30.28it/s]

Reading stations:  99%|█████████▉| 1599/1613 [00:59<00:00, 28.70it/s]

Reading stations:  99%|█████████▉| 1603/1613 [00:59<00:00, 25.50it/s]

Reading stations: 100%|█████████▉| 1607/1613 [00:59<00:00, 27.53it/s]

Reading stations: 100%|█████████▉| 1610/1613 [00:59<00:00, 26.53it/s]

Reading stations: 100%|██████████| 1613/1613 [01:00<00:00, 26.86it/s]

Merged 1,613 stations → 1,610,057 rows × 84 columns in 65 s


In [10]:
aeronet_raw

,site,Date(dd:mm:yyyy),Time(hh:mm:ss),Day_of_Year,AOD_1640nm,AOD_1020nm,AOD_870nm,AOD_865nm,AOD_779nm,AOD_675nm,...,N[340-440_Angstrom_Exponent],N[440-675_Angstrom_Exponent[Polar]],Data_Quality_Level,AERONET_Instrument_Number,AERONET_Site_Name,Site_Latitude(Degrees),Site_Longitude(Degrees),Site_Elevation(m),PI,PI_Email
0,AAOT,04:10:2017,12:00:00,277,NaN,0.169420,NaN,0.227937,0.270857,NaN,...,0,0,lev15,1001,AAOT,45.3139,12.5083,10.0,Giuseppe_Zibordi,giuseppe.zibordi@eoscience.eu
1,AAOT,11:10:2017,12:00:00,284,NaN,0.121093,NaN,0.165021,0.201501,NaN,...,0,0,lev15,1001,AAOT,45.3139,12.5083,10.0,Giuseppe_Zibordi,giuseppe.zibordi@eoscience.eu
2,AAOT,12:10:2017,12:00:00,285,NaN,0.248953,NaN,0.338806,0.409216,NaN,...,0,0,lev15,1001,AAOT,45.3139,12.5083,10.0,Giuseppe_Zibordi,giuseppe.zibordi@eoscience.eu
3,AAOT,13:10:2017,12:00:00,286,NaN,0.270292,NaN,0.344527,0.402378,NaN,...,0,0,lev15,1001,AAOT,45.3139,12.5083,10.0,Giuseppe_Zibordi,giuseppe.zibordi@eoscience.eu
4,AAOT,14:10:2017,12:00:00,287,NaN,0.178536,NaN,0.242100,0.290174,NaN,...,0,0,lev15,1001,AAOT,45.3139,12.5083,10.0,Giuseppe_Zibordi,giuseppe.zibordi@eoscience.eu
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1610052,Zvenigorod,09:10:2021,12:00:00,282,NaN,0.032274,0.038148,NaN,NaN,0.045768,...,15,0,lev15,132,Zvenigorod,55.6950,36.7750,200.0,Pawan_Gupta_and_Elena_Lind,pawan.gupta@nasa.gov_and_elena.lind@nasa.gov
1610053,Zvenigorod,10:10:2021,12:00:00,283,NaN,0.019934,0.026304,NaN,NaN,0.034218,...,33,0,lev15,132,Zvenigorod,55.6950,36.7750,200.0,Pawan_Gupta_and_Elena_Lind,pawan.gupta@nasa.gov_and_elena.lind@nasa.gov
1610054,Zvenigorod,11:10:2021,12:00:00,284,NaN,0.029062,0.039846,NaN,NaN,0.057453,...,6,0,lev15,132,Zvenigorod,55.6950,36.7750,200.0,Pawan_Gupta_and_Elena_Lind,pawan.gupta@nasa.gov_and_elena.lind@nasa.gov
1610055,Zvenigorod,15:10:2021,12:00:00,288,NaN,0.027265,0.037933,NaN,NaN,0.057122,...,1,0,lev15,132,Zvenigorod,55.6950,36.7750,200.0,Pawan_Gupta_and_Elena_Lind,pawan.gupta@nasa.gov_and_elena.lind@nasa.gov


In [11]:
aeronet_raw.columns

Index(['site', 'Date(dd:mm:yyyy)', 'Time(hh:mm:ss)', 'Day_of_Year',
       'AOD_1640nm', 'AOD_1020nm', 'AOD_870nm', 'AOD_865nm', 'AOD_779nm',
       'AOD_675nm', 'AOD_667nm', 'AOD_620nm', 'AOD_560nm', 'AOD_555nm',
       'AOD_551nm', 'AOD_532nm', 'AOD_531nm', 'AOD_510nm', 'AOD_500nm',
       'AOD_490nm', 'AOD_443nm', 'AOD_440nm', 'AOD_412nm', 'AOD_400nm',
       'AOD_380nm', 'AOD_340nm', 'Precipitable_Water(cm)', 'AOD_681nm',
       'AOD_709nm', 'AOD_Empty', 'AOD_Empty.1', 'AOD_Empty.2', 'AOD_Empty.3',
       'AOD_Empty.4', '440-870_Angstrom_Exponent', '380-500_Angstrom_Exponent',
       '440-675_Angstrom_Exponent', '500-870_Angstrom_Exponent',
       '340-440_Angstrom_Exponent', '440-675_Angstrom_Exponent[Polar]',
       'N[AOD_1640nm]', 'N[AOD_1020nm]', 'N[AOD_870nm]', 'N[AOD_865nm]',
       'N[AOD_779nm]', 'N[AOD_675nm]', 'N[AOD_667nm]', 'N[AOD_620nm]',
       'N[AOD_560nm]', 'N[AOD_555nm]', 'N[AOD_551nm]', 'N[AOD_532nm]',
       'N[AOD_531nm]', 'N[AOD_510nm]', 'N[AOD_500nm]', 'N[AO

## 3 · Tidy into an agent-ready table

Exactly the steps of notebook 01, so that the Level 1.5 and Level 2.0 tables have the same columns and can be joined on `AERONET_Site_Name` + `date`:

* keep AERONET's own column names; only replace characters that break pandas / SQL / DuckDB (`Precipitable_Water(cm)` → `Precipitable_Water_cm`, `Site_Latitude(Degrees)` → `Site_Latitude_Degrees`, `N[AOD_500nm]` → `N_AOD_500nm`, `440-870_Angstrom_Exponent` → `Angstrom_Exponent_440_870nm`)
* add a proper `datetime_utc` (and `date`, `year`, `month`) built from AERONET's `dd:mm:yyyy` + `hh:mm:ss` columns
* drop the placeholder `AOD_Empty` columns and any wavelength that is missing at every site
* one consistent column order: site → time → location → AOD → Ångström → water vapour → observation counts → contact


In [12]:
def clean_col(name: str) -> str:
    name = re.sub(r"(\d{3})-(\d{3})_Angstrom_Exponent", r"Angstrom_Exponent_\1_\2nm", name)
    name = name.replace("[Polar]", "_Polar").replace("(%)", "_percent")
    name = re.sub(r"[\(\[]", "_", name)
    name = re.sub(r"[\)\]]", "", name)
    name = re.sub(r"[^0-9A-Za-z_]", "_", name)
    return re.sub(r"_+", "_", name).strip("_")

In [13]:
df = aeronet_raw.copy()

In [14]:
# ── placeholder columns (pandas mangles the duplicated "AOD_Empty" header to AOD_Empty, AOD_Empty.1, …)
df = df.drop(columns=[c for c in df.columns if "AOD_Empty" in c])

In [15]:
# AERONET stamps every daily average at 12:00:00 UTC
df["datetime_utc"] = pd.to_datetime(df["Date(dd:mm:yyyy)"] + " " + df["Time(hh:mm:ss)"], format="%d:%m:%Y %H:%M:%S")
df = df.drop(columns=["Date(dd:mm:yyyy)", "Time(hh:mm:ss)"])

df["date"]  = df["datetime_utc"].dt.normalize()
df["year"]  = df["datetime_utc"].dt.year.astype("int16")
df["month"] = df["datetime_utc"].dt.month.astype("int8")

In [16]:
df.columns = [clean_col(c) for c in df.columns]

In [17]:
# AERONET repeats the site name in every row – make sure it matches the file header, then keep the original column
mismatch = (df["AERONET_Site_Name"] != df["site"]).sum()
print(f"site-name mismatches between header and rows: {mismatch}")
df = df.drop(columns=["site"])

site-name mismatches between header and rows: 0


In [18]:
# ── drop wavelengths that no site ever measured (all-NaN) -----------------------
all_nan = [c for c in df.columns if df[c].isna().all()]
if all_nan:
    print(f"dropping {len(all_nan)} all-NaN columns: {all_nan}")
    df = df.drop(columns=all_nan)

dropping 1 all-NaN columns: ['AOD_531nm']


In [19]:
# … and their (now meaningless) observation-count columns
df = df.drop(columns=[c for c in df.columns if c.startswith("N_") and c[2:] not in df.columns])

In [20]:
# ── dtypes ----------------------------------------------------------------------
for c in df.columns:
    if c.startswith("N_"):
        df[c] = df[c].astype("Int32")
df["AERONET_Site_Name"] = df["AERONET_Site_Name"].astype("string")
for c in ("PI", "PI_Email", "Data_Quality_Level"):
    if c in df.columns:
        df[c] = df[c].astype("string")

In [21]:
# ── column order ----------------------------------------------------------------
def _group(c):
    order = ["AERONET_Site_Name", "datetime_utc", "date", "year", "month", "Day_of_Year",
             "Site_Latitude_Degrees", "Site_Longitude_Degrees", "Site_Elevation_m"]
    if c in order:                       return (0, order.index(c))
    if c.startswith("AOD_"):             return (1, -int(re.search(r"(\d+)nm", c).group(1)))   # long → short wavelength, like AERONET
    if c.startswith("Angstrom_"):        return (2, c)
    if c.startswith("Precipitable_"):    return (3, c)
    if c.startswith("N_"):               return (4, c)
    return (5, c)

In [22]:
df = df[sorted(df.columns, key=_group)]
df = df.sort_values(["AERONET_Site_Name", "datetime_utc"], ignore_index=True)

In [23]:
aeronet = df
del df
print(f"{aeronet.shape[0]:,} rows × {aeronet.shape[1]} columns")
print(list(aeronet.columns))

1,610,057 rows × 73 columns
['AERONET_Site_Name', 'datetime_utc', 'date', 'year', 'month', 'Day_of_Year', 'Site_Latitude_Degrees', 'Site_Longitude_Degrees', 'Site_Elevation_m', 'AOD_1640nm', 'AOD_1020nm', 'AOD_870nm', 'AOD_865nm', 'AOD_779nm', 'AOD_709nm', 'AOD_681nm', 'AOD_675nm', 'AOD_667nm', 'AOD_620nm', 'AOD_560nm', 'AOD_555nm', 'AOD_551nm', 'AOD_532nm', 'AOD_510nm', 'AOD_500nm', 'AOD_490nm', 'AOD_443nm', 'AOD_440nm', 'AOD_412nm', 'AOD_400nm', 'AOD_380nm', 'AOD_340nm', 'Angstrom_Exponent_340_440nm', 'Angstrom_Exponent_380_500nm', 'Angstrom_Exponent_440_675nm', 'Angstrom_Exponent_440_675nm_Polar', 'Angstrom_Exponent_440_870nm', 'Angstrom_Exponent_500_870nm', 'Precipitable_Water_cm', 'N_AOD_1020nm', 'N_AOD_1640nm', 'N_AOD_340nm', 'N_AOD_380nm', 'N_AOD_400nm', 'N_AOD_412nm', 'N_AOD_440nm', 'N_AOD_443nm', 'N_AOD_490nm', 'N_AOD_500nm', 'N_AOD_510nm', 'N_AOD_532nm', 'N_AOD_551nm', 'N_AOD_555nm', 'N_AOD_560nm', 'N_AOD_620nm', 'N_AOD_667nm', 'N_AOD_675nm', 'N_AOD_681nm', 'N_AOD_709nm', 'N_

In [24]:
aeronet

,AERONET_Site_Name,datetime_utc,date,year,month,Day_of_Year,Site_Latitude_Degrees,Site_Longitude_Degrees,Site_Elevation_m,AOD_1640nm,...,N_Angstrom_Exponent_380_500nm,N_Angstrom_Exponent_440_675nm,N_Angstrom_Exponent_440_675nm_Polar,N_Angstrom_Exponent_440_870nm,N_Angstrom_Exponent_500_870nm,N_Precipitable_Water_cm,AERONET_Instrument_Number,Data_Quality_Level,PI,PI_Email
0,AAOT,2017-10-04 12:00:00,2017-10-04,2017,10,277,45.3139,12.5083,10.0,NaN,...,0,25,0,25,0,25,1001,lev15,Giuseppe_Zibordi,giuseppe.zibordi@eoscience.eu
1,AAOT,2017-10-11 12:00:00,2017-10-11,2017,10,284,45.3139,12.5083,10.0,NaN,...,0,20,0,20,0,20,1001,lev15,Giuseppe_Zibordi,giuseppe.zibordi@eoscience.eu
2,AAOT,2017-10-12 12:00:00,2017-10-12,2017,10,285,45.3139,12.5083,10.0,NaN,...,0,33,0,33,0,33,1001,lev15,Giuseppe_Zibordi,giuseppe.zibordi@eoscience.eu
3,AAOT,2017-10-13 12:00:00,2017-10-13,2017,10,286,45.3139,12.5083,10.0,NaN,...,0,7,0,7,0,7,1001,lev15,Giuseppe_Zibordi,giuseppe.zibordi@eoscience.eu
4,AAOT,2017-10-14 12:00:00,2017-10-14,2017,10,287,45.3139,12.5083,10.0,NaN,...,0,12,0,12,0,12,1001,lev15,Giuseppe_Zibordi,giuseppe.zibordi@eoscience.eu
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1610052,Zvenigorod,2021-10-09 12:00:00,2021-10-09,2021,10,282,55.6950,36.7750,200.0,NaN,...,15,15,0,15,15,15,132,lev15,Pawan_Gupta_and_Elena_Lind,pawan.gupta@nasa.gov_and_elena.lind@nasa.gov
1610053,Zvenigorod,2021-10-10 12:00:00,2021-10-10,2021,10,283,55.6950,36.7750,200.0,NaN,...,33,33,0,33,33,33,132,lev15,Pawan_Gupta_and_Elena_Lind,pawan.gupta@nasa.gov_and_elena.lind@nasa.gov
1610054,Zvenigorod,2021-10-11 12:00:00,2021-10-11,2021,10,284,55.6950,36.7750,200.0,NaN,...,6,6,0,6,6,6,132,lev15,Pawan_Gupta_and_Elena_Lind,pawan.gupta@nasa.gov_and_elena.lind@nasa.gov
1610055,Zvenigorod,2021-10-15 12:00:00,2021-10-15,2021,10,288,55.6950,36.7750,200.0,NaN,...,1,1,0,1,1,1,132,lev15,Pawan_Gupta_and_Elena_Lind,pawan.gupta@nasa.gov_and_elena.lind@nasa.gov


## 4 · Sanity checks


In [25]:
print(f"sites        : {aeronet['AERONET_Site_Name'].nunique():,}")
print(f"rows         : {len(aeronet):,}")
print(f"time range   : {aeronet['datetime_utc'].min()}  →  {aeronet['datetime_utc'].max()}")
print(f"duplicates   : {aeronet.duplicated(['AERONET_Site_Name', 'datetime_utc']).sum():,}  (site × datetime)")
print(f"memory       : {aeronet.memory_usage(deep=True).sum() / 1e6:,.0f} MB")

print("\nnon-null fraction per column:")
display((aeronet.notna().mean() * 100).round(1).rename("% non-null").to_frame().T)

display(aeronet.head())

sites        : 1,613
rows         : 1,610,057
time range   : 1993-05-14 12:00:00  →  2026-10-03 12:00:00


duplicates   : 0  (site × datetime)


memory       : 1,165 MB

non-null fraction per column:


,AERONET_Site_Name,datetime_utc,date,year,month,Day_of_Year,Site_Latitude_Degrees,Site_Longitude_Degrees,Site_Elevation_m,AOD_1640nm,...,N_Angstrom_Exponent_380_500nm,N_Angstrom_Exponent_440_675nm,N_Angstrom_Exponent_440_675nm_Polar,N_Angstrom_Exponent_440_870nm,N_Angstrom_Exponent_500_870nm,N_Precipitable_Water_cm,AERONET_Instrument_Number,Data_Quality_Level,PI,PI_Email
% non-null,100.0,100.0,100.0,100.0,100.0,100.0,100.0,100.0,100.0,56.2,...,100.0,100.0,100.0,100.0,100.0,100.0,100.0,100.0,100.0,100.0


,AERONET_Site_Name,datetime_utc,date,year,month,Day_of_Year,Site_Latitude_Degrees,Site_Longitude_Degrees,Site_Elevation_m,AOD_1640nm,...,N_Angstrom_Exponent_380_500nm,N_Angstrom_Exponent_440_675nm,N_Angstrom_Exponent_440_675nm_Polar,N_Angstrom_Exponent_440_870nm,N_Angstrom_Exponent_500_870nm,N_Precipitable_Water_cm,AERONET_Instrument_Number,Data_Quality_Level,PI,PI_Email
0,AAOT,2017-10-04 12:00:00,2017-10-04,2017,10,277,45.3139,12.5083,10.0,NaN,...,0,25,0,25,0,25,1001,lev15,Giuseppe_Zibordi,giuseppe.zibordi@eoscience.eu
1,AAOT,2017-10-11 12:00:00,2017-10-11,2017,10,284,45.3139,12.5083,10.0,NaN,...,0,20,0,20,0,20,1001,lev15,Giuseppe_Zibordi,giuseppe.zibordi@eoscience.eu
2,AAOT,2017-10-12 12:00:00,2017-10-12,2017,10,285,45.3139,12.5083,10.0,NaN,...,0,33,0,33,0,33,1001,lev15,Giuseppe_Zibordi,giuseppe.zibordi@eoscience.eu
3,AAOT,2017-10-13 12:00:00,2017-10-13,2017,10,286,45.3139,12.5083,10.0,NaN,...,0,7,0,7,0,7,1001,lev15,Giuseppe_Zibordi,giuseppe.zibordi@eoscience.eu
4,AAOT,2017-10-14 12:00:00,2017-10-14,2017,10,287,45.3139,12.5083,10.0,NaN,...,0,12,0,12,0,12,1001,lev15,Giuseppe_Zibordi,giuseppe.zibordi@eoscience.eu


## 5 · Export to `DATA/PROCESSED/`

One file: `AERONET_AOD_L15_Daily_V3.parquet` — the full merged table, one row per site × day.


In [26]:
aeronet.to_parquet(OUT_PARQUET, index=False, engine="pyarrow", compression="zstd")
print(f"wrote {OUT_PARQUET}  ({OUT_PARQUET.stat().st_size / 1e6:,.0f} MB, {len(aeronet):,} rows × {aeronet.shape[1]} columns)")

wrote /home/ubuntu/ai_for_aeronet/DATA/PROCESSED/AERONET_AOD_L15_Daily_V3.parquet  (185 MB, 1,610,057 rows × 73 columns)


### Output schema (data dictionary for the agent's skill / system prompt)

The same columns as `AERONET_AOD_L2_Daily_V3.parquet` (notebook 01); the two tables join on `AERONET_Site_Name` + `date`.

| column | original AERONET header | unit | meaning |
|---|---|---|---|
| `AERONET_Site_Name` | `AERONET_Site_Name` | – | site name (string) |
| `datetime_utc`, `date`, `year`, `month` | *derived from* `Date(dd:mm:yyyy)` + `Time(hh:mm:ss)` | UTC | daily averages are stamped 12:00:00 UTC |
| `Day_of_Year` | `Day_of_Year` | day | 1–366 |
| `Site_Latitude_Degrees`, `Site_Longitude_Degrees` | `Site_Latitude(Degrees)`, `Site_Longitude(Degrees)` | decimal degrees | site location (WGS84) |
| `Site_Elevation_m` | `Site_Elevation(m)` | metres | site elevation above sea level |
| `AOD_<λ>nm` | `AOD_<λ>nm` | unitless | aerosol optical depth at wavelength λ nanometres; NaN = not measured / did not pass Level 1.5 screening |
| `Angstrom_Exponent_<λ1>_<λ2>nm` | `<λ1>-<λ2>_Angstrom_Exponent` | unitless | Ångström exponent computed between λ1 and λ2 nm (`_Polar` = polar-instrument variant) |
| `Precipitable_Water_cm` | `Precipitable_Water(cm)` | centimetres | total column water vapour |
| `N_<column>` | `N[<column>]` | count | number of Level 1.5 measurements averaged into that daily value (0 ⇒ value is NaN) |
| `Data_Quality_Level` | `Data_Quality_Level` | – | always `lev15` |
| `AERONET_Instrument_Number` | `AERONET_Instrument_Number` | – | CIMEL sun-photometer id |
| `PI`, `PI_Email` | *file header line 5* | – | site principal investigator (AERONET data-use policy: contact / acknowledge) |

Units reference: <https://aeronet.gsfc.nasa.gov/new_web/units.html>
